File: gai_kappadia.py
import numpy as np
import time

def make_asset_graph(n, tot_assets, integ):
    A_ib = np.random.choice([0, 1], size=(n,n), p=[1 - integ, integ])

    for i in range(n):
        A_ib[i,i] = 0

    for bank in np.where([sum(A_ib[i, 0:n]) == 0 for i in range(n)])[0]:
        if np.size(bank) == 0:
            break
        possible_links = [x for x in range(n) if x != bank]
        A_ib[bank, np.random.choice(possible_links, size = 1)] = 1

    A_ib = tot_assets / np.transpose([sum(A_ib[i, 0:n]) for i in range(n)] * np.transpose(A_ib))
    A_ib = np.where(A_ib == np.inf, 0, A_ib)
    return A_ib

np.sum(A_ib, 1) # sum assets

def is_solvent(bank, A_ib, D, A_M, phi, q):
    (1-phi) * np.sum(A_ib, 1)[bank] + q * np.array(A_M)[bank] - np.sum(A_ib, 0)[bank] - np.array(D)[bank] > 0

def degree(bank, A_ib, direction):
    n = np.size(A_ib, 1)
    if direction == "in":
        return np.sum(A_ib[bank, 0:n] != 0, axis=1)
    else:
        return np.sum(A_ib[0:n, bank] != 0, axis = 0)

def capital_buffer(bank, A_ib, D, A_M):
    return np.sum(A_ib, 1)[bank] + np.array(A_M)[bank] - np.sum(A_ib,0)[bank] - np.array(D)[bank]

# np.sum(A_ib, 0) to liab

def contagion(n, init_b, exp_A_M, exp_A_ib, buffer_rate, link_p):
    A_M = np.random.normal(exp_A,1, n)
    A_ib = make_asset_graph(n, exp_A_ib, link_p)
    D = [exp_A_ib] * n + A_M[0:n] * (1 - buffer_rate) - np.sum(A_ib, 0)[0:n]

    A_ib[0:n, np.random.choice(range(n), 1)] = 0
    n_default = [0]
    n_default = np.append(n_default, sum(capital_buffer(range(n), A_ib, D, A_M) < 0))

    while n_default[-1] != n_default[-2]:
        A_ib[0:n, capital_buffer(range(n), A_ib, D, A_M) < 0] = 0
        n_default = np.append(n_default, sum(capital_buffer(range(n), A_ib, D, A_M) < 0))
    
    return n_default

contagion(1000, 1,80,20,0.04,0.01)

start = time.time()
np.mean([contagion(1000, 1, 80, 20, 0.05, 0.01)[-1] for _ in range(100)])
end = time.time()
print(end - start)

File: gai_kappadia.jl
using Random
using Distributions
using Plots
using LinearAlgebra

# - w funkcji  contagion(): (sum_balance(A_ib, "assets")[1:n] to chyba nie jest potrzebne? Zawsze bedzie tyle samo
# - czy po initial bancruptcy ten wylosowany bank moze wciaz życ? W koncu jedynie usuwamy mu liabilities?
# potem chyba tez - usuwa sie im liab, zostaja A_M i nagle maja dodatni capital
# - chyba trzeba zrobic tak ze pozostale aktywa po defaulcie sa dzielone na wierzycieli

function make_asset_graph(n, tot_assets, integ)
    
    # alternative: A_ib = reduce(hcat,[shuffle(vcat([1 for _ in 1:z], [0 for _ in 1:(n-z)])) for _ in 1:n])'
    A_ib = rand(Bernoulli(integ), n,n)

    for i in 1:n
        A_ib[i,i] = 0
    end

    # if no link then 1 random
    for bank in findall([sum(A_ib[i, 1:n]) != 0 for i in 1:n])
        possible_links = [x for x in 1:n if x != bank]
        A_ib[bank, sample(possible_links, 1)] .= 1
    end 
    A_ib = tot_assets ./ [sum(A_ib[i, 1:n]) for i in 1:n] .* A_ib

    return A_ib
end

function sum_balance(A_ib, class)
    n = size(A_ib)[1]
    if class == "assets"
        return [sum(A_ib[i, 1:n]) for i in 1:n]
    else
        return [sum(A_ib[1:n, i]) for i in 1:n]
    end    
end

function is_solvent(bank, A_ib, D, A_M, ϕ, q)
    (1-ϕ) .* sum_balance(A_ib, "assets")[bank] .+ q .* A_M[bank] .- sum_balance(A_ib, "liab")[bank] - D[bank].> 0
end

function degree(bank, A_ib, direction)
    n = size(A_ib)[1]
    if direction == "in"
        return sum(A_ib[bank, 1:n] .!= 0)
    else
        return sum(A_ib[1:n, bank] .!= 0)
    end
end

function capital_buffer(bank, A_ib, D, A_M)
    return sum_balance(A_ib, "assets")[bank] .+ A_M[bank] - sum_balance(A_ib, "liab")[bank] - D[bank]
end

function contagion(n, init_b, exp_A_M, exp_A_ib, buffer_rate, integ)

    A_M = rand(Normal(exp_A_M,1), n) # illiquid exogenous assets (e.g. mortgage)
    A_ib = make_asset_graph(n, exp_A_ib, integ) # interbank loans from bank i to j

    D = (sum_balance(A_ib, "assets")[1:n] .+ A_M[1:n] .* (1 - buffer_rate)) .- sum_balance(A_ib, "liab")[1:n] # deposits top-up so that capital buffer is 4%
    A_M[sample(1:n, init_b)] .= 0

    n_default = [0]
    n_default = push!(n_default, sum(capital_buffer(collect(1:n), A_ib, D, A_M) .< 0))

    while n_default[end] != n_default[end-1]
        A_ib[1:n, capital_buffer(collect(1:n), A_ib, D, A_M) .< 0] .= 0
        push!(n_default, sum(capital_buffer(collect(1:n), A_ib, D, A_M) .< 0))
    end  

    return n_default
end

function contagion2(n, init_b, exp_A_M, exp_A_ib, buffer_rate, integ)

    A_M = rand(Normal(exp_A_M,0), n) # illiquid exogenous assets (e.g. mortgage)
    A_ib = make_asset_graph(n, exp_A_ib, integ) # interbank loans from bank i to j
    
    D = (sum_balance(A_ib, "assets")[1:n] .+ A_M[1:n] .* (1 - buffer_rate)) .- sum_balance(A_ib, "liab")[1:n] # deposits top-up so that capital buffer is 4%
    A_M[sample(1:n, init_b)] .= 0
    
    defaults_t = [0]
    defaults = []
    defaults = unique(vcat(defaults, findall(capital_buffer(collect(1:n), A_ib, D, A_M) .< 0)))
    push!(defaults_t, length(defaults))
    
    while defaults_t[end] != defaults_t[end-1]
        A_ib[1:n, capital_buffer(collect(1:n), A_ib, D, A_M) .< 0] .= 0
        defaults = unique(vcat(defaults, findall(capital_buffer(collect(1:n), A_ib, D, A_M) .< 0)))
        push!(defaults_t, length(defaults))
    end  

    return defaults_t
end

n = 1000
sys_threshold = 0.05
n_runs = 100
results = []
Random.seed!(1234)

for z in 0:0.0005:0.011
    runs = []
    for contagion in 1:n_runs
        push!(runs, contagion2(n, 1, 80, 20, 0.04, z)[end])     
    end
    systematic = filter(x -> x > n * sys_threshold, runs)
    systematic = length(systematic) == 0 ? 0 : mean(systematic)
    push!(results, [z, systematic, sum(runs .> n * sys_threshold) / n_runs])
    println("calculating contagion with z = ", z, " | extent: ", systematic, " | default prob: ", sum(runs .> n * sys_threshold) / n_runs)
end

results = (reduce(hcat, results[1:end])')
results[:, 2] = results[:, 2] ./ 1000

contagion_plot = plot(results[:,1], results[:,2:3],  label=["Extent of contagion" "Frequency of contagion"])
plot!(legend=:outerbottom, legendcolumns=2)
savefig(contagion_plot, "research_proposal/contagionplot.png")


